jena-arq/src/main/java/org/apache/jena/sparql/algebra/optimize/TransformPathFlattern.java

```diff
@@ -53,9 +58,60 @@ public TransformPathFlattern(PathCompiler pathCompiler)
     @Override
     public Op transform(OpPath opPath)
     {
-        // Flatten down to triples where possible.
-        PathBlock pattern = pathCompiler.reduce(opPath.getTriplePath()) ;
-        // Any generated paths of exactly one to triple; convert to Op.
-        return PathLib.pathToTriples(pattern) ;
+        Op op = transformReduce(opPath, pathCompiler);
+        return op;
     }
-}
+
+    static Op transformReduce(Op op, PathCompiler pathCompiler) {
+        return Transformer.transform(
+                new TransformCopy() {
+                    @Override
+                    public Op transform(OpPath opPath) {
+                        // Flatten down to triples where possible.
+                        PathBlock pattern = pathCompiler.reduce(opPath.getTriplePath());
+                        // Any generated paths of exactly one to triple; convert to Op.
+                        Op op = PathLib.pathToTriples(pattern);
+                        return transformAlt(op, pathCompiler);
+                    }
+                },
+                op);
+    }
+
+    static Op transformAlt(Op op, PathCompiler pathCompiler) {
+        return Transformer.transform(new AltTransformer(pathCompiler), op);
+    }
+
+    private static class AltTransformer extends TransformCopy {
+        private PathCompiler pathCompiler;
+
+        public AltTransformer(PathCompiler pathCompiler) {
+            this.pathCompiler = pathCompiler;
+        }
+
+        @Override
+        public Op transform(OpPath opPath) {
+            TriplePath triplePath = opPath.getTriplePath();
+            Path path = triplePath.getPath();
+            if (path instanceof P_Alt) {
+                Node s = triplePath.getSubject();
```

Comment:
So I don't believe this transformation is semantically safe in all cases.

At a minimum it needs to consider whether either end of the path is a blank node variable as pushing those into a union will make them unique variables relative to the outside of the union and change the semantics of the query.

i.e. a query like `SELECT * WHERE { ?s ex:foo / (ex:a | ex:b) ?o }` can have the first part of the path rewritten to yield `SELECT * WHERE { ?s ex:foo _:bnode . _:bnode (ex:a | ex:b) ?o }`, if you then rewrite the 2nd part into a union you change the semantics -> `SELECT * WHERE { ?s ex:foo _:bnode . { _:bnode ex:a ?o . } UNION { _:bnode ex:b ?o } }`

From the [SPARQL 1.1 specification](https://www.w3.org/TR/sparql11-query/#QSynBlankNodes):

> Blank nodes are indicated by either the label form, such as "_:abc", or the abbreviated form "[]". A blank node that is used in only one place in the query syntax can be indicated with []. A unique blank node will be used to form the triple pattern. Blank node labels are written as "_:abc" for a blank node with label "abc". The same blank node label cannot be used in two different basic graph patterns in the same query.

So that's technically an illegal query you've generated. Evaluation wise those two `_:bnode` variables are now in different scopes and so rather than joining the results of the BGP and the Union you'll be generating a cross-product

Reply:
Hm, The transform happens on the algebra level so all bnodes should have been renamed to variables with appropriate scoping - right?

Reply:
I was using the SPARQL syntax to illustrate the issue but this applies at the algebra nonetheless.

This transform is happening after any scope renaming has happened.  If rewriting an alt to a `union` operator that's also introducing a `join` into the algebra which if there's a blank node variable involved breaks the semantics AFAICT

Reply:
Thank you for the thorough analysis and concern. In my testing, I got the expected results when querying e.g. ` { ?s ex:foo _:bnode . _:bnode (ex:a | ex:b) ?o }` as you suggested. Do you have any additional pointers how to verify the safety of the transform?

Reply:
> In my testing

Which tests are you referring to? There don't seem to be any in the PR. 
`TestTransformPathFlatten`

Reply:
manual testing based on the input provided by @rvesse , I will add some tests to TestTransformPathFlatten